core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

```diff
@@ -582,6 +592,53 @@ private void maybeTerminateChannel() {
     }
   }
 
+  @Override
+  public ConnectivityState getState(boolean requestConnection) {
+    ConnectivityState savedChannelState = channelState;
+    if (requestConnection && savedChannelState == IDLE) {
+      channelExecutor.executeLater(new Runnable() {
+        @Override
+        public void run() {
+          exitIdleMode();
+        }
+      }).drain();
+    }
+    return savedChannelState;
+  }
+
+  private static final class StateChangeCallBack implements Runnable {
+    final ConnectivityState source;
+    final Runnable callback;
+    boolean done;
+
+    StateChangeCallBack(ConnectivityState source, Runnable callBack) {
+      this.source = source;
+      this.callback = callBack;
+    }
+
+    @Override
+    public synchronized void run() {
+      if (!done) {
+        callback.run();
+        done = true;
+      }
+    }
+  }
+
+  @Override
+  public void notifyWhenStateChanged(ConnectivityState source, final Runnable callback) {
+    StateChangeCallBack stateChangCallBack = new StateChangeCallBack(source, callback);
+    if (channelState != source) {
+      executor.execute(stateChangCallBack);
+    } else {
+      stateChangeCallbacks.offer(stateChangCallBack);
+      // check again in case of race with updateChannelState()
```

Comment:
I feel not comfortable with a lock, if an application call notifyWhenStateChanged too frequently by mistake, the contention on the lock may be strong. If we choose to use channelExecutor, no volatile or lock will be needed, which seems fine.